### ISSUE TITLE
"Before Shutdown" hooks for more control over a graceful termination

### ISSUE THREAD

**Issue by stemcc** [loco-rs/loco#760]
In the "after_routes" hook, I am allocating multiple unix file descriptors and spawning multiple external processes that my application needs to interact with over the course of its lifetime.

Since these are external to my application (and external to Rust), I need to terminate them gracefully when my application terminates, or else I can leak file descriptors or cause zombie processes.

Right now to solve this I have to override the serve hook and hook into Axum's graceful shutdown. 

It would be nice not to have to do that and instead have access to a "before_server_shutdown" / "after_server_shutdown" / "main_process_stop" hook where I can gracefully end all the external processes I started.

**Comment by jondot** [comment]
makes sense completely.
im wondering if we want to include a default "best practice" handling or just leave it empty so people override if they want. do you have a best practice shutdown ritual to share?

**Comment by stemcc** [comment]
What I've been doing with Axum is implementing the `with_graceful_shutdown` method on `axum::serve` to call a function that uses `tokio::select!` to match on `signal::ctrl_c()` or `signal::unix::signal(signal::unix::SignalKind::terminate())` to call the  actors I manage my external processes with to shut everything down. Once all those handles return then I know it's safe to shutdown.

I imagine another way would be to run all the cleanup code at the very end of the `main()` function after the axum server returns (which is how I've done it in Actix before). 

I've used Sanic in Python for a similar project and it has also got the notion of hooks (it calls signals) for hooking into an app's lifecycle, and there I used its `server.shutdown.before` and `server.shutdown.after` signals for ensuring everything was fully cleaned up. 

I've used [Tokio's Graceful Shutdown](https://tokio.rs/tokio/topics/shutdown) docs and Axum's [graceful shutdown docs](https://docs.rs/axum/latest/axum/serve/struct.WithGracefulShutdown.html) and examples [1](https://github.com/tokio-rs/axum/blob/main/examples/graceful-shutdown/src/main.rs) and [2](https://github.com/tokio-rs/axum/blob/main/examples/tls-graceful-shutdown/src/main.rs) to reach my current implementation. I'm not 100% certain it's the best way to do this but I wanted to keep as much in Axum as I possibily could.
